modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -98,265 +101,291 @@
+     */
+    private synchronized void initChannelHolders(boolean force) {
+        // enable parallel threads to schedule new init of channel holders
+        scheduledChannelsReinit.set(false);
+
+        if (!force && channels.get() != null)
+            return;
+
+        List<InetSocketAddress> resolvedAddrs = parseAddresses(clientCfg.getAddresses());
+
+        List<ClientChannelHolder> holders = Optional.ofNullable(channels.get()).orElse(new ArrayList<>());
+
+        // addr -> (holder, delete)
+        Map<InetSocketAddress, T2<ClientChannelHolder, Boolean>> addrs = holders.stream()
+            .collect(Collectors.toMap(
+                c -> c.chCfg.getAddress(),
+                c -> new T2<>(c, null)
+        ));
+
+        // mark for delete addrs that aren't provided by clientConfig now
+        addrs.keySet()
+            .stream()
+            .filter(addr -> !resolvedAddrs.contains(addr))
+            .forEach(addr -> addrs.get(addr).setValue(true));
+
+        // create new holders for new addrs
+        resolvedAddrs.stream()
+            .filter(addr -> !addrs.containsKey(addr))
+            .forEach(addr -> {
+                ClientChannelHolder hld = new ClientChannelHolder(
+                    new ClientChannelConfiguration(clientCfg, addr), chFactory, onChannelCreate, onChannelClose);
+
+                addrs.put(addr, new T2<>(hld, false));
+            });
+
+        if (!stopInitCondition()) {
+            List<ClientChannelHolder> list = new ArrayList<>();
+            // The variable holds a new index of default channel after topology change.
+            // Suppose that reuse of the channel is better than open new connection.
+            int dfltChannelIdx = -1;
+
+            ClientChannelHolder currHolder = null;
+            if (curChIdx != -1)
+                currHolder = channels.get().get(curChIdx);
+
+            for (T2<ClientChannelHolder, Boolean> t : addrs.values()) {
+                ClientChannelHolder hld = t.get1();
+                Boolean markForDelete = t.get2();
+
+                if (markForDelete == null) {
+                    // this channel is still in use
+                    list.add(hld);
+                    if (hld == currHolder)
+                        dfltChannelIdx = list.size() - 1;
 
-        affinityCtx = new ClientCacheAffinityContext(binary);
+                }
+                else if (markForDelete) {
+                    // this holder should be deleted now
+                    nodeChannels.values().remove(hld);
+                    hld.close();
+                }
+                else {
+                    // this channel is new
+                    list.add(hld);
+                }
+            }
 
-        ClientConnectionException lastEx = null;
+            if (dfltChannelIdx == -1)
+                dfltChannelIdx = new Random().nextInt(list.size());
 
-        for (int i = 0; i < channels.length; i++) {
+            curChannelsGuard.writeLock().lock();
```

Comment:
Do we need a lock around AtomicReference usage?

Reply:
This lock is around two variables that must be in sync: channels (AtomicReference) and curChIdx.

Reply:
Then maybe `AtomicReference` is not required? Furthermore, this method is `synchronized` - overall rather hard to understand.

Reply:
There are different goals for those sync operations:
1. Synchronized method guarantees that only one thread is initializing holders. This is required to limit requests to the AddressFinder, and avoid do the same work for prepare addresses multiple times;
2. AtomicReference is required as list of addresses are changed due to the preparation - merging new addresses, removing obsolete addresses. This performed in the "thin-client-channel#" thread.
3. The same time channels and CurChIdx are used in non-synchronized methods in user threads. There should be a guarantee for consistency between those variables. So this a reason to use the guard.

Reply:
Agree with Pavel, we don't need `AtomicReference`, it's too verbose and we don't use any atomic features (no CAS operations, only get and set). Can be replaced with volatile variable (List or array).